daemon/logs.go

```diff
@@ -10,40 +10,52 @@ import (
 	"sync"
 
 	"github.com/Sirupsen/logrus"
-	"github.com/docker/docker/engine"
 	"github.com/docker/docker/pkg/jsonlog"
+	"github.com/docker/docker/pkg/stdcopy"
 	"github.com/docker/docker/pkg/tailfile"
 	"github.com/docker/docker/pkg/timeutils"
+	"github.com/docker/docker/pkg/version"
 )
 
-func (daemon *Daemon) ContainerLogs(job *engine.Job) error {
-	if len(job.Args) != 1 {
-		return fmt.Errorf("Usage: %s CONTAINER\n", job.Name)
-	}
+type ContainerLogsConfig struct {
+	Follow, Timestamps   bool
+	Tail                 string
+	UseStdout, UseStderr bool
+	OutStream            io.Writer
+	ApiVersion           version.Version
+}
 
+func (daemon *Daemon) ContainerLogs(name string, config *ContainerLogsConfig) error {
 	var (
-		name   = job.Args[0]
-		stdout = job.GetenvBool("stdout")
-		stderr = job.GetenvBool("stderr")
-		tail   = job.Getenv("tail")
-		follow = job.GetenvBool("follow")
-		times  = job.GetenvBool("timestamps")
 		lines  = -1
 		format string
 	)
-	if !(stdout || stderr) {
+	if !(config.UseStdout || config.UseStderr) {
 		return fmt.Errorf("You must choose at least one stream")
 	}
-	if times {
+	if config.Timestamps {
 		format = timeutils.RFC3339NanoFixed
 	}
-	if tail == "" {
-		tail = "all"
+	if config.Tail == "" {
+		config.Tail = "all"
 	}
+
 	container, err := daemon.Get(name)
 	if err != nil {
 		return err
 	}
+
+	var (
+		outStream = config.OutStream
+		errStream io.Writer
+	)
+	if !container.Config.Tty && config.ApiVersion.GreaterThanOrEqualTo("1.6") {
```

Comment:
I'm pretty sure that version check is not necessary. I blindly copied it from attach, but `/logs/` endpoint isn't even exists before "1.6"